pkg/controller/jobs/kubeflow/trainjob/trainjob_controller.go

```diff
@@ -0,0 +1,249 @@
+		SetupWebhook:      SetupTrainJobWebhook,
+		JobType:           &kftrainerapi.TrainJob{},
+		AddToScheme:       kftrainerapi.AddToScheme,
+		MultiKueueAdapter: &multiKueueAdapter{},
+	}))
+}
+
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainjobs/finalizers,verbs=get;update
+
+type trainJobReconciler struct {
+	jr     *jobframework.JobReconciler
+	client client.Client
+}
+
+var reconciler trainJobReconciler
+var _ jobframework.JobReconcilerInterface = (*trainJobReconciler)(nil)
+
+func NewReconciler(client client.Client, eventRecorder record.EventRecorder, opts ...jobframework.Option) jobframework.JobReconcilerInterface {
+	reconciler = trainJobReconciler{
+		jr:     jobframework.NewReconciler(client, eventRecorder, opts...),
+		client: client,
+	}
+	return &reconciler
+}
+
+func (r *trainJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	jobSet := jobsetapi.JobSet{}
+	err := r.client.Get(ctx, req.NamespacedName, &jobSet)
+	if err != nil {
+		// we only reconcile the TrainJob when its child JobSet is available
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	return r.jr.ReconcileGenericJob(ctx, req, &TrainJob{})
+}
+
+func (r *trainJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	b := ctrl.NewControllerManagedBy(mgr).
+		For(&kftrainerapi.TrainJob{}).Owns(&kueue.Workload{}).Owns(&jobsetapi.JobSet{})
+	return b.Complete(r)
+}
+
+type TrainJob kftrainerapi.TrainJob
+
+var _ jobframework.GenericJob = (*TrainJob)(nil)
+var _ jobframework.JobWithReclaimablePods = (*TrainJob)(nil)
+var _ jobframework.JobWithManagedBy = (*TrainJob)(nil)
+
+func NewJob() jobframework.GenericJob {
+	return &TrainJob{}
+}
+
+func fromObject(obj runtime.Object) *TrainJob {
+	return (*TrainJob)(obj.(*kftrainerapi.TrainJob))
+}
+
+func (t *TrainJob) Object() client.Object {
+	return (*kftrainerapi.TrainJob)(t)
+}
+
+func (t *TrainJob) IsSuspended() bool {
+	return ptr.Deref(t.Spec.Suspend, false)
+}
+
+func (t *TrainJob) IsActive() bool {
+	for i := range t.Status.JobsStatus {
+		if t.Status.JobsStatus[i].Active > 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func (t *TrainJob) Suspend() {
+	t.Spec.Suspend = ptr.To(true)
+}
+
+func (t *TrainJob) Unsuspend() {
+	t.Spec.Suspend = ptr.To(false)
+}
+
+func (t *TrainJob) GVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (t *TrainJob) PodLabelSelector() string {
+	return fmt.Sprintf("%s=%s", jobsetapi.JobSetNameKey, t.Name)
+}
+
+func getChildJobSet(t *TrainJob) (*jobsetapi.JobSet, error) {
+	jobSet := jobsetapi.JobSet{}
+	// TODO: Ideally we should be using the parent context here
+	err := reconciler.client.Get(context.Background(), types.NamespacedName{Name: t.Name, Namespace: t.Namespace}, &jobSet)
+	return &jobSet, err
```

Comment:
if err then this should return nil, err.

Reply:
Yes, but looking at all usages it is anyway ignored on errors. We can address in a follow up. I also know that @kaisoz is still trying to refactor the code to eliminate the lookup to JobSet, so this part of code would be refactored anyway.